spec/requests/saved_pages_request_spec.rb

```diff
@@ -0,0 +1,169 @@
+RSpec.describe "SavedPages", type: :request do
+  context "when receiving an unauthenticated request" do
+    it "returns unauthorised for GET /api/saved_pages" do
+      get saved_pages_path
+
+      expect(response).to have_http_status(:unauthorized)
+    end
+
+    it "returns unauthorised for POST /api/saved_pages" do
+      post saved_pages_path({ page_path: "/my-saved-page-path" })
+
+      expect(response).to have_http_status(:unauthorized)
+    end
+
+    it "returns unauthorised for DELETE /api/saved_pages" do
+      delete saved_page_path({ page_path: "/my-saved-page-path" })
+
+      expect(response).to have_http_status(:unauthorized)
+    end
+
+    it "returns unauthorised for GET /api/saved_pages/:page_path " do
+      get saved_page_path("/my-saved-page-path")
+
+      expect(response).to have_http_status(:unauthorized)
+    end
+  end
+
+  context "when receiving an authenticated request" do
+    let(:oidc_client) { instance_double(OidcClient) }
+    let(:headers) { { "Content-Type" => "application/json", "GOVUK-Account-Session" => session_identifier } }
+    let(:session_identifier) { placeholder_govuk_account_session(user_id: user.sub) }
+    let(:user) { FactoryBot.create(:oidc_user) }
+
+    before do
+      stub_oidc_discovery
+
+      allow(OidcClient).to receive(:new).and_return(oidc_client)
+      allow(oidc_client).to receive(:tokens!).and_return({ access_token: "access-token", refresh_token: "refresh-token" })
+    end
+
+    describe "GET /api/saved_pages" do
+      it "returns an empty array if there are no saved pages" do
+        get saved_pages_path, headers: headers
+
+        expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body)["saved_pages"]).to eq([])
+      end
+
+      it "returns an array of saved_pages if they exist" do
+        list = FactoryBot.create_list(:saved_page, 2, oidc_user_id: user.id)
+        expected_response = list.map { |page| { "page_path" => page.page_path } }
+
+        get saved_pages_path, headers: headers
+
+        expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body)["saved_pages"]).to eq(expected_response)
+      end
+    end
+
+    describe "POST /api/saved_pages" do
+      it "returns a page path hash if the page persists correctly" do
+        post saved_pages_path, headers: headers, params: { page_path: "/page-path/1" }.to_json
+
+        expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body)["saved_page"]).to eq({ "page_path" => "/page-path/1" })
+      end
+
+      it "returns status 200 and upserts the record if the page already exists" do
+        FactoryBot.create(:saved_page, oidc_user_id: user.id, page_path: "/page-path/1")
+        post saved_pages_path, headers: headers, params: { page_path: "/page-path/1" }.to_json
+
+        expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body)["saved_page"]).to eq({ "page_path" => "/page-path/1" })
+      end
```

Comment:
We're not testing that a record gets created in the first test and that it doesn't in the second test.

Reply:
[Done](https://github.com/alphagov/account-api/pull/69/commits/0c4ad32ab6c9b7f9d1be36f59645c6689a838a7b#diff-e494be0ec09eb92d5782bcb4d976f0a1c24e9f3874d5b2a1e980ac902d51eda7R56-R60)